sound/soc/sof/intel/hda-dai.c

```diff
@@ -359,65 +357,35 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
 	rtd = asoc_substream_to_rtd(substream);
 	codec_dai = asoc_rtd_to_codec(rtd, 0);
 
-	switch (cmd) {
-	case SNDRV_PCM_TRIGGER_START:
-	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
-		snd_hdac_ext_stream_start(hext_stream);
-		if (pipeline->state != SOF_IPC4_PIPE_PAUSED) {
-			ret = sof_ipc4_set_pipeline_state(sdev, pipe_widget->instance_id,
-							  SOF_IPC4_PIPE_PAUSED);
-			if (ret < 0)
-				return ret;
-			pipeline->state = SOF_IPC4_PIPE_PAUSED;
-		}
-
-		ret = sof_ipc4_set_pipeline_state(sdev, pipe_widget->instance_id,
-						  SOF_IPC4_PIPE_RUNNING);
+	if (ops && ops->dma_ops && ops->dma_ops->pre_trigger) {
+		ret = ops->dma_ops->pre_trigger(sdev, dai, substream, cmd);
 		if (ret < 0)
 			return ret;
-		pipeline->state = SOF_IPC4_PIPE_RUNNING;
-		break;
-	case SNDRV_PCM_TRIGGER_SUSPEND:
-	case SNDRV_PCM_TRIGGER_STOP:
-	{
-		ret = sof_ipc4_set_pipeline_state(sdev, pipe_widget->instance_id,
-						  SOF_IPC4_PIPE_PAUSED);
+	}
+
+	if (ops && ops->dma_ops && ops->dma_ops->trigger) {
+		ret = ops->dma_ops->trigger(sdev, dai, substream, cmd);
 		if (ret < 0)
 			return ret;
+	}
 
-		pipeline->state = SOF_IPC4_PIPE_PAUSED;
-
-		snd_hdac_ext_stream_clear(hext_stream);
-
-		ret = sof_ipc4_set_pipeline_state(sdev, pipe_widget->instance_id,
-						  SOF_IPC4_PIPE_RESET);
+	if (ops && ops->dma_ops && ops->dma_ops->post_trigger) {
+		ret = ops->dma_ops->post_trigger(sdev, dai, substream, cmd);
```

Comment:
Ok, so here at least dma_ops pre_trigger(), trigger(), and post_trigger() are called in sequence, without anything happening in between. I trust this not the situation everywhere. It would be really nice to have a big picture about in which order and where these different ops are called in different situations, or at least some examples about the sequence.